polars/polars-lazy/polars-plan/src/logical_plan/projection.rs

```diff
@@ -208,15 +208,12 @@ fn expand_dtypes(
     dtypes: &[DataType],
     exclude: &[Arc<str>],
 ) -> PolarsResult<()> {
-    for dtype in dtypes {
-        for field in schema.iter_fields().filter(|f| f.data_type() == dtype) {
+    for field in schema.iter_fields() {
```

Comment:
I think we can add a `dtype_processed_count` and break once we processed all dtypes. Given that there are users with millions of columns this `N x D` iteration might become expensive.

Could we also add a comment that we loop over the schema in order to maintain the schema order independently of the `dtype` order given.

Reply:
Actually... If I write it a little more cleanly as...
```rust
for field in schema.iter_fields().filter(|f| dtypes.contains(&f.dtype)) {
```
...vs the original...
```rust
for dtype in dtypes {
    for field in schema.iter_fields().filter(|f| f.data_type() == dtype) { 
```
...it would seem that the ideal optimisation would probably be to hold `dtypes` in a set instead of a vec, to make the `contains` check cheaper when we have a larger number of dtypes? 🤔

Then it's just a single iteration over the fields with a hash comparison, instead of potentially `n dtypes` loops over all fields (what it's currently doing).

Reply:
The linear search probably is cheaper as we only have a few dtypes.

Reply:
Was wondering if dtypes were sprrtable, then we could do a binary search, heh. Probably overkill. In which case it's probably faster now than it was before (where the inner check filters against all the fields each time). I just updated with the slightly cleaner filter construction.